### PR TITLE
Fix seed to generate deterministic samples

### LINKED ISSUE #690: Improve `sample()` method
### Problem Description

Let's make sampling more user friendly. We can create multiple methods for different user needs.

The `sample()` method should only be used to sample entire rows (no conditions).

### Expected behavior

Parameters:
- `num_rows` is a required param
- `randomize_samples` will determine whether or not there should be a fixed seed (default: True)
- Remove all existing params

```python
# works with any tabular model
from sdv.tabular import CTGAN
model = CTGAN()
model.fit(data)

# creates 1000 rows. Different rows created every time
model.sample(num_rows=1000)

# Fix a seed so that this call always creates rows in the same order
model.sample(num_rows=1000, randomize_samples=False)

```

## Error Handling

Guidance if the user does not specify `num_rows`.
```python
>>> model.sample()
Error: You must specify the number of rows to sample (eg. num_rows=100).
```

Guidance if the user passes in the `conditions` argument.

```python
>>> model.sample(num_rows=100, conditions={'age': 45})
Error: This method does not support the conditions parameter. Please create Condition objects
and pass them into the sample_conditions method. See User Guide or API for more details.
```

### PR DESCRIPTION
**Pull request by katxiao**
Resolves #690

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -18,8 +18,8 @@
     "numpy>=1.20.0,<2;python_version>='3.7'",
     'pandas>=1.1.3,<2',
     'tqdm>=4.15,<5',
-    'copulas>=0.6.0,<0.7',
-    'ctgan>=0.5.0,<0.6',
+    'copulas>=0.6.1.dev0,<0.7',
```
Comment (katxiao) [inline setup.py:21]: I'll update these version numbers once `copulas` and `ctgan` are released.

File: sdv/tabular/base.py [hunk 2]
```diff
@@ -390,6 +391,20 @@ def _conditionally_sample_rows(self, dataframe, max_retries, max_rows_multiplier
 
         return sampled_rows
 
+    def _randomize_samples(self, randomize_samples):
```
Comment (fealho) [inline sdv/tabular/base.py:394]: If it's easy, I'd add a test case for this as well.

File: sdv/tabular/copulas.py [hunk 3]
```diff
@@ -318,6 +318,15 @@ def _sample(self, num_rows, conditions=None):
         """
         return self._model.sample(num_rows, conditions=conditions)
 
+    def _set_random_state(self, random_state):
+        """Set the random state of the model's random number generator.
+
+        Args:
+            random_state (int, np.random.RandomState, or None):
```
Comment (fealho) [inline sdv/tabular/copulas.py:325]: What's the expected behavior if it's passed None? Maybe add a test for it, if the behavior is any different than passing an integer.

**Review by fealho — APPROVED** [review]
Looks good to me, just some small details.

File: sdv/tabular/copulas.py (hunk 3 above)
Comment (katxiao) [reply sdv/tabular/copulas.py:325]: It should set the underlying model's random state to None. Not sure if we should test it here since it's the internal model behavior (e.g. within CTGAN and copulas)

File: sdv/tabular/ctgan.py [hunk 4]
```diff
@@ -79,6 +79,15 @@ def _sample(self, num_rows, conditions=None):
 
         raise NotImplementedError(f"{self._MODEL_CLASS} doesn't support conditional sampling.")
 
+    def _set_random_state(self, random_state):
+        """Set the random state of the model's random number generator.
+
+        Args:
+            random_state (int, np.random.RandomState, or None):
```
Comment (csala) [inline sdv/tabular/ctgan.py:86]: Wouldn't this be `tuple[np.random.RandomState, torch.Generator]` instead of just `np.random.RandomState`?

File: sdv/tabular/base.py [hunk 5]
```diff
@@ -390,6 +391,19 @@ def _conditionally_sample_rows(self, dataframe, max_retries, max_rows_multiplier
...
+        """Randomize the samples according to user input.
+
+        If ``randomize_samples`` is false, fix the seed that the random number generator
+        uses in the underlying models.
+
+        Args:
+            randomize_samples (bool):
+                Whether or not to randomize the generated samples.
+        """
+        if randomize_samples:
```
Comment (csala) [inline sdv/tabular/base.py:404]: Shouldn't this be `if not`?

File: sdv/tabular/base.py [hunk 6]
```diff
@@ -390,6 +391,19 @@ def _conditionally_sample_rows(self, dataframe, max_retries, max_rows_multiplier
 
         return sampled_rows
 
+    def _randomize_samples(self, randomize_samples):
```
Comment (csala) [inline sdv/tabular/base.py:394]: I don't see this used anywhere apart from the unit tests.
On the other hand, what happens if `randomize_samples` is `True`?
Should we set the seed to None?

Comment (katxiao) [reply sdv/tabular/base.py:427]: @csala Yes, if `randomize_samples` is true, we should set the seed to `None`. I updated the code to call the `_randomize_samples` method, and I also added an integration test to test that setting the seed to None works as expected.
